tests/helpers/deviceappproxy/apduabstract.py

```diff
@@ -0,0 +1,129 @@
+            return data
+        if isinstance(data, int):
+            return bytes([cast(int, data)])
+        if isinstance(data, str):
+            return bytes.fromhex(data)
+        raise TypeError(f"{data} cannot be converted to bytes")
+
+    @property
+    def apdus(self) -> Optional[ApduDict]:
+        return ApduSet._apdus if len(ApduSet._apdus.keys()) > 0 else None
+
+    @apdus.setter
+    def apdus(self, new_apdus: ApduDict, overwrite: bool = False) -> None:
+        """Sets a new CApsu internal dictionary if it wasn't set at instanciation time,
+        unless overwrite is True."""
+        if not self.apdus or overwrite is True:
+            if not isinstance(new_apdus, dict):
+                raise ValueError("Attribute newApdus must be a dictionary containing CApdu "
+                                 "instances as values")
+            ApduSet._apdus = new_apdus
+
+    def apdu(self, name: str,
+             p1: Optional[AnyStr] = None,
+             p2: Optional[AnyStr] = None,
+             data: Optional[List[AnyStr]] = None,
+             le: Optional[AnyStr] = None) -> Tuple[bytes, List[Optional[bytes]]]:
+        """Returns the raw bytes for the C-APDU header requested by name.
+        """
+
+        def _bytesbuf(apdu: CApdu, apdu_keys: List[str]) -> bytes:
+            """Concatenates all @apdu attributes whose names are provided in @apdu_keys,
+            into a single byte buffer."""
+            return b''.join(self._bytes(getattr(apdu, k)) if k in apdu.__dict__ else self._bytes(k) for k in apdu_keys)
+
+        if not self.apdus:
+            raise ValueError("ApduSet object is empty! Provide an ApduDict either at instanciation"
+                             " or with the 'apdus' attribute.")
+        if name not in self.apdus:
+            raise KeyError(f"{name} APDU is not supported by this instance")
+        # Compose APDU depending on its type into a byte buffer
+        self.set_params(key=name, p1=p1, p2=p2, data=data, le=le)
+        # Determine APDU type
+        apdu_is_in_only_or_inout: bool = self._apdus[name].typ == CApdu.Type.IN \
+                                         or self._apdus[name].typ == CApdu.Type.INOUT
+        apdu_is_out_only: bool = self._apdus[name].typ == CApdu.Type.OUT
+        # Return the C-APDU header with correct Lc
+        return (
+            _bytesbuf(
+                self.apdus[name],
+                ['cla', 'ins', 'p1', 'p2', 'lc' if apdu_is_in_only_or_inout else 'le' if apdu_is_out_only else '00']
+            ),
+            self.apdus[name].data
+        )
+
+    def __setitem__(self, key: str, value: CApdu) -> None:
+        """Change an existing APDU or add a new one to the APDU dict
+        """
+        if not isinstance(value, CApdu):
+            raise ValueError(f"Syntax '{self.__class__.__name__}[{key}] = value' "
+                             f"only accept CApdu instances as value")
+        self.apdus[key] = value
```

Comment:
this is suprising: ApduSet has half of the interface of a dict (so name is confusing) and this setter does not actually seem to be used anyway.

Maybe we can get rid of this?

Reply:
- Overloaded __setitem__() special func: it used to be used but i've made many design changes since so... After commenting it out, tests still continue to pass ==> I've removed it. 

- Name confusion: you are right, this class is wrapper + some useful methods around a dict, it is not a set. ==> Renamed type `ApduDict` to `ApduDictType` and class `ApduSet` to `ApduDict`.

Reply:
👍